tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -3705,8 +3889,8 @@ private[optimizer] object OptimizerCore {
     }
 
     def apply(binding: PreTransBinding,
-        result: PreTransLocalDef): PreTransform = {
-      if (result.contains(binding.localDef)) {
+        result: PreTransResult): PreTransform = {
+      if (result.contains(binding.localDef) || binding.isAlreadyUsed) {
```

Comment:
Ah, yes. I almost thought there was a pre-existing major bug when I had to add that one.

The thing is, before, the only possible `result` was `PreTransLocalDef`. A `PreTransLocalDef` can only ever contain virtual references to local variables, and so `contains` catches all possible references to local variables defined by `binding`.

But now, `result` can also be a `PreTransBinaryOp`, for example. And one of the arguments could be a `PreTransTree`, which itself contains somewhere a `VarRef` to the variable defined by `binding`. That would happen, for example, if an operand was an `Apply(..., List(VarRef(...)))`, since the `Apply` cannot be virtualized, its arguments had to be eagerly be forced, and a `VarRef` was produced (setting `used.value = true` in the process). After that, the top-level `PreTransBinaryOp` does `contains(binding.localDef)`, but it used the binding nevertheless.